### PR TITLE
Feature state reduction using conservation laws

### PR DESCRIPTION
**Pull request by FFroehlich**

### CODE REVIEW

File: python/amici/ode_export.py [hunk 1]
```diff
@@ -1494,18 +1494,19 @@ def _totalDerivative(self, name, eq, chainvar, var,
             else:
                 variables[var]['sym'] = self.eq(varname)
 
+        self._eqs[name] = \
+            self._eqs[name] = copy.deepcopy(variables['dydz']['sym'])
```
Comment (FFroehlich) [inline python/amici/ode_export.py]: @dweindl note that a you need to deepcopy here, did not spot this when I reviewed your code.

Comment (dweindl) [reply python/amici/ode_export.py]: Thx. Will merge fix with #500.

File: python/amici/ode_export.py [hunk 2]
```diff
@@ -765,18 +911,68 @@ def add_component(self, component):
...
+            state in the model
+
+        """
+        try:
+            ix = [
+                s.get_id()
+                for s in self._states
+            ].index(state)
+        except ValueError:
+            raise Exception(f'Speciefied state {state} was not found in the '
```
Comment (dweindl) [inline python/amici/ode_export.py]:
```suggestion
            raise Exception(f'Specified state {state} was not found in the '
```

File: python/amici/ode_export.py [hunk 3]
```diff
@@ -765,18 +911,68 @@ def add_component(self, component):
...
+                            f'model states.')
+
+        total_abundace = sp.Symbol(f'tcl_{self._states[ix].get_id()}')
```
Comment (dweindl) [inline python/amici/ode_export.py]:
```suggestion
        total_abundance = sp.Symbol(f'tcl_{self._states[ix].get_id()}')
```

File: python/amici/ode_export.py [hunk 4]
```diff
@@ -765,18 +911,68 @@ def add_component(self, component):
...
+
+        state_id = self._states[ix].get_id()
+
+        self.add_component(
+            Expression(
+                state_id,
+                f'cl_{state_id}',
+                total_abundace - law
```
Comment (dweindl) [inline python/amici/ode_export.py]:
```suggestion
                total_abundance - law
```

File: python/amici/ode_export.py [hunk 5]
```diff
@@ -765,18 +911,68 @@ def add_component(self, component):
...
+            )
+        )
+
+        self.add_component(
+            ConservationLaw(
+                total_abundace,
```
Comment (dweindl) [inline python/amici/ode_export.py]:
```suggestion
                total_abundance,
```

File: python/amici/ode_export.py [hunk 6]
```diff
@@ -765,18 +911,68 @@ def add_component(self, component):
...
+                f'total_{state_id}',
+                state_id + law
+            )
+        )
+
+        self._states[ix].set_conservation_law(
+            total_abundace - law
```
Comment (dweindl) [inline python/amici/ode_export.py]:
```suggestion
            total_abundance - law
```

File: python/amici/ode_export.py [hunk 7]
```diff
@@ -786,9 +982,35 @@ def nx(self):
...
         """
         return len(self.sym('x'))
 
+    def ncl(self):
+        """Number of conservation laws.
+
+        Arguments:
+
+        Returns:
+        number of state variable symbols
```
Comment (dweindl) [inline python/amici/ode_export.py]:
```suggestion
        number of conservation laws
```

File: python/amici/ode_export.py [hunk 8]
```diff
@@ -2146,4 +2525,86 @@ def sanitize_basic_sympy(basic):
...
+        # not as functions etc
+        locals = {
+            fs.name: sp.Symbol(fs.name)
+            for fs in list(basic.expr_free_symbols)
+            if isinstance(fs, pysb.core.Component)
+        }
+        return sp.sympify(str(basic), locals=locals)
+
+
+def get_function_defition(fun, name):
```
Comment (dweindl) [inline python/amici/ode_export.py]:
```suggestion
def get_function_definition(fun, name):
```

File: python/amici/ode_export.py [hunk 9]
```diff
@@ -2146,4 +2525,86 @@ def sanitize_basic_sympy(basic):
...
+        signature: function signature @type str
+
+    Returns:
+    string that can be used to construct function calls with the same
+    variable names and ordering as in the function signature
+
+    Raises:
+
+    """
+    typedefs = [
```
Comment (dweindl) [inline python/amici/ode_export.py]: Inconsistent final space in contained strings. Intended?

File: src/backwardproblem.cpp [hunk 10]
```diff
@@ -185,11 +185,11 @@ void BackwardProblem::handleEventB(int iroot) {
  * @param[in] it index of data point @type int
  */
 void BackwardProblem::handleDataPointB(int it) {
-    for (int ix = 0; ix < model->nxtrue; ix++) {
+    for (int ix = 0; ix < model->nxtrue_solver; ix++) {
         for (int iJ = 0; iJ < model->nJ; iJ++)
-            // we only need the 1:nxtrue slice here!
-            xB[ix + iJ * model->nxtrue] +=
-                dJydx[iJ + ( ix + it * model->nx ) * model->nJ];
+            // we only need the 1:nxtrue_cl slice here!
```
Comment (dweindl) [inline src/backwardproblem.cpp]: Not sure I understand how the comment relates to the code

File: src/model.cpp [hunk 11]
```diff
@@ -549,14 +545,16 @@ void Model::setTimepoints(const std::vector<realtype> &ts) {
...
     return stateIsNonNegative;
 }
 
 void Model::setStateIsNonNegative(std::vector<bool> const& stateIsNonNegative) {
-    if (stateIsNonNegative.size() != static_cast<unsigned long>(nx))
-        throw AmiException("Dimension of input stateIsNonNegative (%d) does not agree with number of state variables (%d)",stateIsNonNegative.size(),nx);
+    if(nx_solver != nx_rdata)
+        throw AmiException("Nonnegative states are not supported whith conservation laws enabled");
+    if (stateIsNonNegative.size() != static_cast<unsigned long>(nx_solver))
+        throw AmiException("Dimension of input stateIsNonNegative (%d) does not agree with number of state variables (%d)",stateIsNonNegative.size(),nx_solver);
```
Comment (dweindl) [inline src/model.cpp]:
```suggestion
        throw AmiException("Dimension of input stateIsNonNegative (%u) does not agree with number of state variables (%d)",stateIsNonNegative.size(),nx_solver);
```

File: src/model.cpp [hunk 12]
```diff
@@ -589,7 +587,10 @@ std::vector<realtype> const& Model::getInitialStates() const {
 }
 
 void Model::setInitialStates(const std::vector<realtype> &x0) {
-    if(x0.size() != (unsigned) nx && x0.size() != 0)
+    if(nx_solver != nx_rdata)
+        throw AmiException("Custom initial conditions are not supported whith conservation laws enabled");
+    // TODO: remove this exception and compute total_cl according to provided x_rdata
```
Comment (dweindl) [inline src/model.cpp]: Create issue?

File: python/amici/ode_export.py (hunk 9 above)
Comment (FFroehlich) [reply python/amici/ode_export.py]: In parts, will add explanatory comments (the `*` is for pointers). Whitespace after `int` was indeed missing.

File: src/model.cpp (hunk 12 above)
Comment (FFroehlich) [reply src/model.cpp]: This is actually not too difficult to implement now. would probably require some additional tests (which could be added to model presimulation)

**Pull request merged**